### PR TITLE
Extracted Elasticsearch indexer common functionality

### PR DESCRIPTION
**Pull request by slovit**

### CODE REVIEW

File: dcc-common-es/pom.xml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-core</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-databind</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>com.fasterxml.jackson.dataformat</groupId>
```
Comment (btiernay) [inline dcc-common-es/pom.xml:59]: Can we make this an optional maven dependency? Or is this required?

File: dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/DocumentWriterFactory.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	 * Creates the client settings.
+	 * <p>
+	 * 
+	 * @see http://www.elasticsearch.org/guide/en/elasticsearch/client/java-api/current/client.htmls
+	 */
+	private static Builder createSettings(boolean sniff) {
+		return ImmutableSettings.settingsBuilder()
+
+		// Increase the ping timeout from the 5s (default) to something larger to prevent transient
+		// NoNodeAvailableExceptions
```
Comment (btiernay) [inline dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/DocumentWriterFactory.java]: Odd formatting

File: dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/DocumentWriterFactory.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	@SneakyThrows
+	private static int getPort(String esUri) {
+		return new URI(esUri).getPort();
+	}
+
+	/**
+	 * Creates the client settings.
+	 * <p>
```
Comment (btiernay) [inline dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/DocumentWriterFactory.java]: Remove.

File: dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/DocumentWriterFactory.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+				.put("client.transport.nodes_sampler_interval", "10s")
+
+				// Enable / disable the client to sniff the rest of the cluster, and add those into its list of machines to use.
+				// In this case, note that the IP addresses used will be the ones that the other nodes were started with (the
+				// "publish" address)
+				.put("client.transport.sniff", sniff);
+	}
+
+	private static String createWriterId() {
+		val id = RANDOM_GENERATOR.nextInt(Integer.MAX_VALUE);
```
Comment (btiernay) [inline dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/DocumentWriterFactory.java]: Formatter seems to be using tabs. Can we reformat with the standard spaces? Thanks.

File: dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/model/EsDocument.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+import lombok.NonNull;
+import lombok.Value;
+
+import org.icgc.dcc.dcc.common.es.impl.EsDocumentType;
+
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+@Value
+public class EsDocument {
```
Comment (btiernay) [inline dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/model/EsDocument.java]: Can we change this to `Document` and use `ReleaseDocument` in `dcc-release`?

File: dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/model/EsDocument.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+	@NonNull
+	String id;
+	@NonNull
+	ObjectNode source;
+	@NonNull
+	EsDocumentType type;
```
Comment (btiernay) [inline dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/model/EsDocument.java]: Same question here.

File: dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/impl/EsDocumentWriter.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+import org.icgc.dcc.dcc.common.es.model.EsDocument;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectWriter;
+
+/**
+ * Output destination for {@link DefaultDocument} instances to be written.
+ */
+@Slf4j
+public class EsDocumentWriter implements DocumentWriter {
```
Comment (btiernay) [inline dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/impl/EsDocumentWriter.java]: Why do we need `DocumentWriter` and `EsDocumentWriter` in the `dcc-common-es` module? It is ES after all.

File: dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/impl/BulkProcessorListener.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+  @Override
+  public void afterBulk(long executionId, BulkRequest request, Throwable failure) {
+    log.debug("[{}] Received unsuccessful response for request {}", writerId, executionId);
+    indexingState.enableCheckClusterState();
+    log.debug("{}", indexingState);
+
+    log.warn("[{}] Encountered exception during bulk load: {}", writerId, getExceptionMessage(failure));
+    retryRequest(executionId, request);
+    log.debug("{}", indexingState);
+
```
Comment (btiernay) [inline dcc-common-es/src/main/java/org/icgc/dcc/dcc/common/es/impl/BulkProcessorListener.java]: Remove.

File: dcc-common-es/README.md [hunk 9]
```diff
@@ -0,0 +1,12 @@
+ICGC DCC - Common Elasticsearch
+===
```
Comment (btiernay) [inline dcc-common-es/README.md]: Please switch to `#` style

File: dcc-common-es/README.md [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+Elasticsearch tools for DCC projects
+
+Build
+---
+
+In the repository's root execute from the command line:
+
+        mvn -am -pl dcc-common-es package
```
Comment (btiernay) [inline dcc-common-es/README.md]: And ````shell` style

File: dcc-common-es/pom.xml (hunk 1 above)
Comment (slovit) [reply dcc-common-es/pom.xml:59]: Required

Comment (slovit) [reply dcc-common-es/pom.xml:59]: Required

File: dcc-common-es/README.md (hunk 9 above)
Comment (slovit) [reply dcc-common-es/README.md]: You first :) 
https://raw.githubusercontent.com/icgc-dcc/dcc-common/develop/dcc-common-ega/README.md

**Pull request merged**